### PR TITLE
CSApply: Handle unbound references to methods found via dynamic lookup

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
Builds on #41849

```swift
class Class {
  @objc func f() {}
}

let x: (AnyObject) -> (() -> Void)? = AnyObject.f
```

### CODE REVIEW

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

File: lib/SIL/Verifier/SILVerifier.cpp [hunk 1]
```diff
@@ -3378,36 +3379,67 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
 
     assert(!methodTy->isCoroutine());
 
-    // Map interface types to archetypes.
-    if (auto *env = F.getModule().Types.getConstantGenericEnvironment(method)) {
-      auto subs = env->getForwardingSubstitutionMap();
-      methodTy = methodTy->substGenericArgs(F.getModule(), subs,
-                                            F.getTypeExpansionContext());
+    // If the method comes from a protocol, substitute 'Self' type of 'self' at
+    // the call site. Otherise, map interface types to archetypes.
```
Comment (LucianoPAlmeida) [inline lib/SIL/Verifier/SILVerifier.cpp:3383]: Typo `Otherise` -> `Otherwise`?

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
cc @xedin

File: lib/Sema/CSApply.cpp [hunk 2]
```diff
@@ -1452,24 +1489,28 @@ namespace {
...
+        // For a partial application of a protocol method, open the existential
+        // inside the curry thunk as well. This reduces abstraction and
+        // post-factum function type conversions, and results in better SILGen.
         //
-        // However, a partial application of a class method on a subclass
-        // existential does need to open the existential, so that it can be
-        // upcast to the appropriate class reference type.
-        if (!isPartialApplication || !containerTy->hasOpenedExistential()) {
+        // For a partial application of a class method, however, we always want
+        // the thunk to accept a class to avoid potential abstraction, so the
```
Comment (slavapestov) [inline lib/Sema/CSApply.cpp:1508]: What is "abstraction" in this context?

File: test/Sema/dynamic_self_implicit_conversions.swift [hunk 3]
```diff
@@ -1,5 +1,6 @@
 // RUN: %target-swift-frontend -typecheck -dump-ast %s | %FileCheck %s
 
+// FIXME: Make this a SILGen test instead.
```
Comment (slavapestov) [inline test/Sema/dynamic_self_implicit_conversions.swift:3]: Is there more to this than just moving the test to test/SILGen and redoing the CHECK: lines to look for SIL instructions?

**Review by slavapestov — APPROVED** [review]

**Comment by slavapestov** [comment]
Do you mind adding CHANGELOG entries for this and https://github.com/apple/swift/pull/41849?

**Comment by AnthonyLatsis** [comment]
#41849 made it to 5.7. Does it make sense to ship this alongside?

**Comment by slavapestov** [comment]
I think it's pretty low risk. Can you tag @hborla on the cherry-pick?

**Review by xedin — APPROVED** [review]

**Comment by AnthonyLatsis** [comment]
@swift-ci  please test source compatibility

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Comment by AnthonyLatsis** [comment]
Windows failure unrelated, source compat shows a UPASS from some earlier change.

**Pull request merged**